### ISSUE TITLE
allow().to receive doesn't respect order

### ISSUE THREAD

**Issue by rwz** [rspec/rspec-mocks#997]
```ruby
require "rspec/autorun"

class MyClass
  def initialize(client)
    @client = client
  end

  def fetch_results
    [].tap do |results|
      results << @client.get_value while results.length < 3
    end
  end
end

describe MyClass do
  let(:client){ double("Client") }
  subject{ described_class.new(client) }

  it "works" do
    expect(client).to receive(:get_value).and_return(1).ordered
    expect(client).to receive(:get_value).and_return(2).ordered
    expect(client).to receive(:get_value).and_return(3).ordered

    expect(subject.fetch_results).to eq([1,2,3])
  end

  it "should work, but fails" do
    allow(client).to receive(:get_value).and_return(1).ordered
    allow(client).to receive(:get_value).and_return(2).ordered
    allow(client).to receive(:get_value).and_return(3).ordered

    expect(subject.fetch_results).to eq([1,2,3])
  end
end
```

This is expected to pass in both cases but only actually passes with `expect` syntax.

```
Failures:

  1) MyClass should work, but fails
     Failure/Error: expect(subject.fetch_results).to eq([1,2,3])

       expected: [1, 2, 3]
            got: [3, 3, 3]

       (compared using ==)
     # run.rb:32:in `block (2 levels) in <main>'

Finished in 0.01525 seconds (files took 0.09345 seconds to load)
2 examples, 1 failure

Failed examples:

rspec 'run.rb[1:2]' # MyClass should work, but fails
```

**Comment by myronmarston** [comment]
If you want to return different sequential values for multiple calls, just pass multiple values to `and_return`:

https://relishapp.com/rspec/rspec-mocks/v/3-3/docs/configuring-responses/returning-a-value#specify-different-return-values-for-multiple-calls

The problem in your example is that stubs _replace_ earlier stubs rather than specifying responses for later calls.  This is intended to support cases like:

``` ruby
RSpec.describe do
  before do
    allow(Collaborator).to receive(:foo).and_return(:generic_response)
  end

  it "does x" do
    # ...
  end

  it "does y" do
    # ...
  end

  context "with a more specific response" do
    before do
      allow(Collaborator).to receive(:foo).and_return(:specific response)
    end

    it "does z" do
      # ...
    end
  end
end
```

Here the second `allow` overrides what was set as the response in the first `before` hook.  We can't make `allow` additive like you thought it was and support that.

Message expectations (e.g. via `expect`) are additive because each one specifies something that _must_ be satisfied for the example to pass, so we have to keep each one.

Do you have any suggestions for how we can improve our docs to make this more clear?

**Comment by JonRowe** [comment]
We should probably stop people calling `ordered` on `allow`...

**Comment by myronmarston** [comment]
Yep.

**Comment by JonRowe** [comment]
I'll do that tonight

**Comment by JonRowe** [comment]
Closed by #1000
